pkg/endpoint/redirect_test.go

```diff
@@ -478,6 +482,315 @@ func (s *RedirectSuite) TestRedirectWithDeny(c *check.C) {
+		c.Fatal("desired policy map does not equal expected map:\n",
+			ep.desiredPolicy.GetPolicyMap().Diff(c.T, expected))
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cmp := completion.NewWaitGroup(ctx)
+
+	realizedRedirects := ep.GetRealizedRedirects()
+	desiredRedirects, err, finalizeFunc, revertFunc := ep.addNewRedirects(cmp)
+	c.Assert(err, check.IsNil)
+	finalizeFunc()
+
+	// Check that all redirects have been created.
+	c.Assert(desiredRedirects["12345:ingress:TCP:80:/cec2/listener2"], check.Equals, crd2Port)
+	c.Assert(desiredRedirects["12345:ingress:TCP:80:/cec1/listener1"], check.Equals, crd1Port)
+	c.Assert(len(desiredRedirects), check.Equals, 2)
+
+	expected2 := policy.NewMapState(map[policy.Key]policy.MapStateEntry{
+		mapKeyAllowAllE: {
+			DerivedFromRules: labels.LabelArrayList{AllowAnyEgressLabels},
+		},
+		mapKeyFooL7: {
+			ProxyPort:        crd2Port,
+			Listener:         "/cec2/listener2",
+			DerivedFromRules: labels.LabelArrayList{lblsL4AllowListener1, lblsL4AllowPort80, lblsL4L7AllowListener2Priority1},
+		},
+		mapKeyAllL7: {
+			DerivedFromRules: labels.LabelArrayList{lblsL4AllowListener1, lblsL4AllowPort80, lblsL4L7AllowListener2Priority1},
+		},
+	})
+	if !ep.desiredPolicy.GetPolicyMap().Equals(expected2) {
+		c.Fatal("desired policy map does not equal expected map:\n",
+			ep.desiredPolicy.GetPolicyMap().Diff(c.T, expected2))
+	}
+
+	// Keep only desired redirects
+	ep.removeOldRedirects(desiredRedirects, realizedRedirects, cmp)
+
+	// Check that the redirect is still realized
+	c.Assert(len(desiredRedirects), check.Equals, 2)
+	c.Assert(ep.desiredPolicy.GetPolicyMap().Len(), check.Equals, 3)
+
+	// Pretend that something failed and revert the changes
+	revertFunc()
+
+	// Check that the state before addRedirects is restored
+	if !ep.desiredPolicy.GetPolicyMap().Equals(expected) {
+		c.Fatal("desired policy map does not equal expected map:\n",
+			ep.desiredPolicy.GetPolicyMap().Diff(c.T, expected))
+	}
+	c.Assert(ep.desiredPolicy.GetPolicyMap().Len(), check.Equals, 2)
+}
+
+func (s *RedirectSuite) TestRedirectWithEqualPriority(c *check.C) {
+	// Setup dependencies for endpoint.
+	kvstore.SetupDummy(c, "etcd")
+
+	oldPolicyEnable := policy.GetPolicyEnabled()
+	defer policy.SetPolicyEnabled(oldPolicyEnable)
+	policy.SetPolicyEnabled(option.DefaultEnforcement)
+
+	idAllocatorOwner := &DummyIdentityAllocatorOwner{}
+
+	mgr := cache.NewCachingIdentityAllocator(idAllocatorOwner)
+	<-mgr.InitIdentityAllocator(nil)
+	defer mgr.Close()
+
+	identityCache := cache.IdentityCache{
+		identity.NumericIdentity(identityFoo): labelsFoo,
+		identity.NumericIdentity(identityBar): labelsBar,
+	}
+
+	do := &DummyOwner{
+		repo: policy.NewPolicyRepository(mgr, identityCache, nil, nil),
+	}
+	do.repo.GetSelectorCache().SetLocalIdentityNotifier(testidentity.NewDummyIdentityNotifier())
+	identitymanager.Subscribe(do.repo)
+	defer identitymanager.RemoveAll()
```

Comment:
(There's a bit of code here that seems like boilerplate across multiple tests, does it differ? Is there an opportunity for a helper here or is it awkward to do so?)

Reply:
Good call. Added new `SetUpTest()` `TearDownTest()` and helpers for Endpoint and policy rules to remove this and more of the boilerplate.